### PR TITLE
DRM Support

### PR DESCRIPTION
**Pull request by tinusneethling**
Generic interfaces for DRM support.

First implementation of ClearKey for Andoird.

DRM Support #468

### CODE REVIEW

**Comment by ryanheise** [comment]
Thanks for your work. It looks like I forgot to push my dev branch with 0.9.5. Can you do a rebase to include that?

File: just_audio_platform_interface/lib/just_audio_platform_interface.dart [hunk 1]
```diff
@@ -1008,69 +1008,77 @@ abstract class IndexedAudioSourceMessage extends AudioSourceMessage {
...
-    required this.uri,
-    this.headers,
-    dynamic tag,
-  }) : super(id: id, tag: tag);
+  UriAudioSourceMessage(
+      {required String id,
+      required this.uri,
+      this.headers,
+      dynamic tag,
+      this.drm})
```
Comment (ryanheise) [inline just_audio_platform_interface/lib/just_audio_platform_interface.dart:1018]: Should be a trailing comma here to preserve formatting.

File: just_audio_platform_interface/lib/just_audio_platform_interface.dart [hunk 2]
```diff
@@ -1008,69 +1008,77 @@ abstract class IndexedAudioSourceMessage extends AudioSourceMessage {
...
-    required String uri,
-    Map<String, String>? headers,
-    dynamic tag,
-  }) : super(id: id, uri: uri, headers: headers, tag: tag);
+  ProgressiveAudioSourceMessage(
+      {required String id,
+      required String uri,
+      Map<String, String>? headers,
+      dynamic tag,
+      Map<String, String>? drm})
```
Comment (ryanheise) [inline just_audio_platform_interface/lib/just_audio_platform_interface.dart:1030]: Missing trailing comma here.

File: just_audio_platform_interface/lib/just_audio_platform_interface.dart [hunk 3]
```diff
@@ -1008,69 +1008,77 @@ abstract class IndexedAudioSourceMessage extends AudioSourceMessage {
...
-    required String uri,
-    Map<String, String>? headers,
-    dynamic tag,
-  }) : super(id: id, uri: uri, headers: headers, tag: tag);
+  DashAudioSourceMessage(
+      {required String id,
+      required String uri,
+      Map<String, String>? headers,
+      dynamic tag,
+      Map<String, String>? drm})
```
Comment (ryanheise) [inline just_audio_platform_interface/lib/just_audio_platform_interface.dart:1051]: Missing trailing comma.

File: just_audio_platform_interface/lib/just_audio_platform_interface.dart [hunk 4]
```diff
@@ -1008,69 +1008,77 @@ abstract class IndexedAudioSourceMessage extends AudioSourceMessage {
...
-    required String uri,
-    Map<String, String>? headers,
-    dynamic tag,
-  }) : super(id: id, uri: uri, headers: headers, tag: tag);
+  HlsAudioSourceMessage(
+      {required String id,
+      required String uri,
+      Map<String, String>? headers,
+      dynamic tag,
+      Map<String, String>? drm})
```
Comment (ryanheise) [inline just_audio_platform_interface/lib/just_audio_platform_interface.dart:1072]: Missing trailing comma

File: just_audio/lib/just_audio.dart [hunk 5]
```diff
@@ -2155,12 +2165,19 @@ class ProgressiveAudioSource extends UriAudioSource {
...
+      : super(uri, headers: headers, tag: tag, duration: duration, drm: drm);
 
   @override
   AudioSourceMessage _toMessage() => DashAudioSourceMessage(
-      id: _id, uri: _effectiveUri.toString(), headers: headers, tag: tag);
+      id: _id,
+      uri: _effectiveUri.toString(),
+      headers: headers,
+      tag: tag,
+      drm: drm);
```
Comment (ryanheise) [inline just_audio/lib/just_audio.dart:2180]: Missing trailing comma.

File: just_audio/lib/just_audio.dart [hunk 6]
```diff
@@ -2155,12 +2165,19 @@ class ProgressiveAudioSource extends UriAudioSource {
 /// your device to forward HTTP requests with headers included.
 class DashAudioSource extends UriAudioSource {
   DashAudioSource(Uri uri,
-      {Map<String, String>? headers, dynamic tag, Duration? duration})
-      : super(uri, headers: headers, tag: tag, duration: duration);
+      {Map<String, String>? headers,
+      dynamic tag,
+      Duration? duration,
+      Map<String, String>? drm})
```
Comment (ryanheise) [inline just_audio/lib/just_audio.dart:2171]: Missing trailing comma.

File: just_audio/lib/just_audio.dart [hunk 7]
```diff
@@ -2131,12 +2134,19 @@ abstract class UriAudioSource extends IndexedAudioSource {
...
+      : super(uri, headers: headers, tag: tag, duration: duration, drm: drm);
 
   @override
   AudioSourceMessage _toMessage() => ProgressiveAudioSourceMessage(
-      id: _id, uri: _effectiveUri.toString(), headers: headers, tag: tag);
+      id: _id,
+      uri: _effectiveUri.toString(),
+      headers: headers,
+      tag: tag,
+      drm: drm);
```
Comment (ryanheise) [inline just_audio/lib/just_audio.dart:2149]: Missing trailing comma.

File: just_audio/lib/just_audio.dart [hunk 8]
```diff
@@ -2131,12 +2134,19 @@ abstract class UriAudioSource extends IndexedAudioSource {
 /// your device to forward HTTP requests with headers included.
 class ProgressiveAudioSource extends UriAudioSource {
   ProgressiveAudioSource(Uri uri,
-      {Map<String, String>? headers, dynamic tag, Duration? duration})
-      : super(uri, headers: headers, tag: tag, duration: duration);
+      {Map<String, String>? headers,
+      dynamic tag,
+      Duration? duration,
+      Map<String, String>? drm})
```
Comment (ryanheise) [inline just_audio/lib/just_audio.dart:2140]: Missing trailing comma.

File: just_audio/lib/just_audio.dart [hunk 9]
```diff
@@ -2043,9 +2044,11 @@ abstract class IndexedAudioSource extends AudioSource {
 abstract class UriAudioSource extends IndexedAudioSource {
   final Uri uri;
   final Map<String, String>? headers;
+  final Map<String, String>? drm;
   Uri? _overrideUri;
 
-  UriAudioSource(this.uri, {this.headers, dynamic tag, Duration? duration})
+  UriAudioSource(this.uri,
+      {this.headers, dynamic tag, Duration? duration, this.drm})
```
Comment (ryanheise) [inline just_audio/lib/just_audio.dart:2051]: Trailing comma

File: just_audio/lib/just_audio.dart [hunk 10]
```diff
@@ -2043,9 +2044,11 @@ abstract class IndexedAudioSource extends AudioSource {
 abstract class UriAudioSource extends IndexedAudioSource {
   final Uri uri;
   final Map<String, String>? headers;
+  final Map<String, String>? drm;
```
Comment (ryanheise) [inline just_audio/lib/just_audio.dart:2047]: I think this will need a doc comment.

File: just_audio/lib/just_audio.dart [hunk 11]
```diff
@@ -1978,16 +1979,16 @@ abstract class AudioSource {
   /// If headers are set, just_audio will create a cleartext local HTTP proxy on
   /// your device to forward HTTP requests with headers included.
```
Comment (ryanheise) [inline just_audio/lib/just_audio.dart:1980]: Add mention of the drm parameter here.

File: just_audio/android/src/main/java/com/ryanheise/just_audio/AudioPlayer.java [hunk 12]
```diff
@@ -576,52 +580,63 @@ private MediaSource getAudioSource(final Object json) {
...
+        if (map.get("drm") != null) {
+            Map<?, ?> drm = (Map<?, ?>) map.get("drm");
+            if (drm.get("clearKey") != null) {
+                drmSessionManager = new DefaultDrmSessionManager.Builder()
+                        .setUuidAndExoMediaDrmProvider(C.CLEARKEY_UUID, FrameworkMediaDrm.DEFAULT_PROVIDER).
+                                build(new LocalMediaDrmCallback(((String) drm.get("clearKey")).getBytes()));
+            }
+        }
+
+        switch ((String) map.get("type")) {
```
Comment (ryanheise) [inline just_audio/android/src/main/java/com/ryanheise/just_audio/AudioPlayer.java:596]: I'm not sure if I want to change the project's formatting style, but if we do it, it should be first proposed in an issue and done as a separate pull request, otherwise it adds changes to the diff that are harder to review.

File: just_audio/android/src/main/java/com/ryanheise/just_audio/AudioPlayer.java [hunk 13]
```diff
@@ -576,52 +580,63 @@ private MediaSource getAudioSource(final Object json) {
...
-        default:
-            throw new IllegalArgumentException("Unknown AudioSource type: " + map.get("type"));
+        Map<?, ?> map = (Map<?, ?>) json;
+        String id = (String) map.get("id");
+
+        DrmSessionManager drmSessionManager = null;
+        if (map.get("drm") != null) {
+            Map<?, ?> drm = (Map<?, ?>) map.get("drm");
+            if (drm.get("clearKey") != null) {
+                drmSessionManager = new DefaultDrmSessionManager.Builder()
```
Comment (ryanheise) [inline just_audio/android/src/main/java/com/ryanheise/just_audio/AudioPlayer.java:590]: In future, how would this approach be extended to support a custom DrmSessionManager?

**Comment by ryanheise** [comment]
Hi @tinusneethling 

Thanks for contributing! I've just finished reviewing and left some comments (mostly superficial) but the main point to think about is future extensibility. One future requirement may be to support custom DRM session managers on the Android side (and whatever the equivalent is on the iOS side if any). For example, to implement offline playback, which is a common requirement:

https://github.com/google/ExoPlayer/issues/949

At this stage, implementing that is not necessary, just future-proofing the design in a way that's extensible down the track, particularly the platform interface which should always be backwards compatible.

**Comment by ryanheise** [comment]
Offline DRM playback on the iOS side:

https://developer.apple.com/videos/play/wwdc2020/10655/

(Just adding this link here to collect all relevant info in one place, even if it's not the immediate goal to implement it.)

File: just_audio/android/src/main/java/com/ryanheise/just_audio/AudioPlayer.java (hunk 13 above)
Comment (tinusneethling) [reply just_audio/android/src/main/java/com/ryanheise/just_audio/AudioPlayer.java:590]: New methos should be added to the Dart helper calls(DRMHelper) to create a Map with the required key value pairs for the DRM. The platform specific code should then check if those properties are available and instantiate the DRMManager accordingly.(Android)

File: just_audio/android/src/main/java/com/ryanheise/just_audio/AudioPlayer.java (hunk 12 above)
Comment (tinusneethling) [reply just_audio/android/src/main/java/com/ryanheise/just_audio/AudioPlayer.java:596]: Changes not related to formatting but setDrmSessionManager(drmSessionManager) added to the various MediaSource factories.

File: just_audio/android/src/main/java/com/ryanheise/just_audio/AudioPlayer.java (hunk 13 above)
Comment (ryanheise) [reply just_audio/android/src/main/java/com/ryanheise/just_audio/AudioPlayer.java:590]: I understand the design, but a custom DrmCessionManager should be able to implement callbacks with custom logic, particularly to support use cases like offline playback.

File: just_audio/android/src/main/java/com/ryanheise/just_audio/AudioPlayer.java (hunk 12 above)
Comment (ryanheise) [reply just_audio/android/src/main/java/com/ryanheise/just_audio/AudioPlayer.java:596]: Would you be able to undo the formatting changes to this file?

File: just_audio/CHANGELOG.md [hunk 14]
```diff
@@ -1,3 +1,8 @@
+## 0.9.5
```
Comment (ryanheise) [inline just_audio/CHANGELOG.md:1]: This part of the change log should already be in `dev` so I'm not sure why it's showing up in the diff.

**Comment by ryanheise** [comment]
Closing due to #477

File: just_audio/android/src/main/java/com/ryanheise/just_audio/AudioPlayer.java (hunk 13 above)
Comment (tinusneethling) [reply just_audio/android/src/main/java/com/ryanheise/just_audio/AudioPlayer.java:590]: The DRM Clearkey is support for offline playback. The key/kid pair is provided to the DRM manager and if media asset contains the KID then corresponding key is taken.(call back happens natively)

For callbacks this is done in online mode where the URL is called to relevant server. The idea would be the specify URL(if possible) and token information in the drm map which would be provided to the DRMSesssion manager.
